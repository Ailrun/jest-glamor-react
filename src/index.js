const matcher = require('./matcher')
const serializer = require('./serializer')

module.exports = {
  matcher,
  serializer,
}
