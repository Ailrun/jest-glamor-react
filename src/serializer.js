const css = require('css')
const {styleSheet} = require('glamor')

const getClassNames = (node, classNames) => {
  if (node.children && node.children.reduce) {
    classNames = node.children.reduce(
      (acc, child) => acc.concat(getClassNames(child, acc)),
      classNames,
    )
  }
  const className = node.props && (node.props.className || node.props.class)
  if (className) {
    return classNames.concat(className.toString().split(' '))
  }
  return []
}

const getStyles = classNames => {
  const styles = styleSheet.tags
    .map(tag => /* istanbul ignore next */ tag.textContent || '')
    .join('\n')
  const ast = css.parse(styles)
  ast.stylesheet.rules = ast.stylesheet.rules.filter(
    rule =>
      rule.type === 'rule' &&
      classNames.includes(rule.selectors[0].substring(1)),
  )

  const ret = css.stringify(ast)
  return ret
}

const serializer = {
  test(val) {
    return !val.withStyles && val.$$typeof === Symbol.for('react.test.json')
  },

  print(val, print) {
    const classNames = getClassNames(val, [])
    const styles = getStyles(classNames)
    val.withStyles = true
    const printedVal = print(val)
    if (styles) {
      return `${styles}\n\n${printedVal}`
    } else {
      return printedVal
    }
  },
}

module.exports = serializer
