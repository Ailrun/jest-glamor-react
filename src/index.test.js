import React from 'react'
import renderer from 'react-test-renderer'
import * as glamor from 'glamor'
import * as enzyme from 'enzyme'
import toJson from 'enzyme-to-json'
import {matcher, serializer} from '../src'

expect.addSnapshotSerializer(serializer)
expect.extend(matcher)

function Wrapper(props) {
  const className = glamor.css({
    padding: '4em',
    background: 'papayawhip',
  })
  return <section className={`${className}`} {...props} />
}

function Title(props) {
  const className = glamor.css({
    fontSize: '1.5em',
    textAlign: 'center',
    color: 'palevioletred',
  })
  return <h1 className={`${className}`} {...props} />
}

test('react-test-renderer', () => {
  const tree = renderer
    .create(
      <Wrapper>
        <Title>Hello World, this is my first glamor styled component!</Title>
      </Wrapper>,
    )
    .toJSON()

  expect(tree).toMatchSnapshotWithGlamor()
})

test('enzyme', () => {
  const ui = (
    <Wrapper>
      <Title>Hello World, this is my first glamor styled component!</Title>
    </Wrapper>
  )

  const enzymeMethods = ['shallow', 'mount', 'render']
  enzymeMethods.forEach(method => {
    const tree = enzyme[method](ui)
    expect(toJson(tree)).toMatchSnapshotWithGlamor(`enzyme.${method}`)
  })
})
