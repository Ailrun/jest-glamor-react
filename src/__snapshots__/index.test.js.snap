// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enzyme.mount 1`] = `
<Wrapper>
  <section
    className="css-1otybxc"
  >
    <Title>
      <h1
        className="css-1tnuino"
      >
        Hello World, this is my first glamor styled component!
      </h1>
    </Title>
  </section>
</Wrapper>
`;

exports[`enzyme.render 1`] = `
.css-1otybxc,
[data-css-1otybxc] {
  padding: 4em;
  background: papayawhip;
}

.css-1tnuino,
[data-css-1tnuino] {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  class="css-1otybxc"
>
  <h1
    class="css-1tnuino"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;

exports[`enzyme.shallow 1`] = `
.css-1otybxc,
[data-css-1otybxc] {
  padding: 4em;
  background: papayawhip;
}

<section
  className="css-1otybxc"
>
  <Title>
    Hello World, this is my first glamor styled component!
  </Title>
</section>
`;

exports[`react-test-renderer 1`] = `
.css-1otybxc,
[data-css-1otybxc] {
  padding: 4em;
  background: papayawhip;
}

.css-1tnuino,
[data-css-1tnuino] {
  font-size: 1.5em;
  text-align: center;
  color: palevioletred;
}

<section
  className="css-1otybxc"
>
  <h1
    className="css-1tnuino"
  >
    Hello World, this is my first glamor styled component!
  </h1>
</section>
`;
